### PR TITLE
Options to disable certain messages

### LINKED ISSUE #990: [Request] Options to disable certain messages
ZDoom and derivatives have a setting for "Minimum message level", meaning that certain kinds of messages can be hidden without outright disabling all messages.

I think that this could be a nice addition for Woof, albeit providing separate settings for each kind of message. By "kind" I mean categories like:

- Item pickup messages;
- Cheat messages;
- Toggle messages (like "Mouselook On/Off");
- Warning messages (like the "Unknown Thing" warning).

Those are the ones I remember off the top of my head.

Implementing this would surely mean looking through the code for all instances of messages and adding conditions to them. If you'd like to go ahead with it, I can try to implement it myself and make a PR for it.

### PR DESCRIPTION
**Pull request by MrAlaux**
That's a lot of messages... I hope I didn't miss anything.

Fixes #990

### CODE REVIEW

File: src/m_menu.c [hunk 1]
```diff
@@ -76,6 +76,7 @@ int mouseSensitivity_horiz2; // [FG] strafe
 int mouseSensitivity_vert2; // [FG] look
 
 int showMessages;    // Show messages has default, 0 = off, 1 = on
+int hideMessages;
```
Comment (MrAlaux) [inline src/m_menu.c:79]: Not too sure about this variable name, as it might get confusing given that we already have `showMessages`. Suggestions?

Comment (fabiangreffrath) [reply src/m_menu.c:79]: `hide_message_type` maybe?

Comment (MrAlaux) [reply src/m_menu.c:79]: Sounds good.

**Comment by kraflab** [comment]
I'm not sure I get the motivation behind this change in its current form. If we want to have support for hiding message categories, this doesn't have much coverage. Basically, if you're going to go through a change that edits every line that adds a message, why not go all the way? I would expect this feature to involve a menu page with all the category options with toggles so people can customize things.

**Comment by MrAlaux** [comment]
> I would expect this feature to involve a menu page with all the category options with toggles so people can customize things.

That was my original idea, actually, but Fabian doesn't seem to be very fond of having too many menu items - which is fine, really. Still, in its current form, I think this covers a fair chunk of (mildly annoying) messages in practice.

File: src/m_menu.c [hunk 2]
```diff
@@ -4522,8 +4524,15 @@ static void M_UpdateMultiLineMsgItem(void)
   DISABLE_ITEM(!message_list, mess_settings1[mess_lines]);
 }
 
+static const char *hide_message_strings[] = {
+  "none", "toggles", "pickups", "both", NULL
```
Comment (rfomin) [inline src/m_menu.c:4528]: I suggest two options on the menu:
`toggle messages: yes/no`
`pickup messages: yes/no`

The current multi-item option is a bit confusing to me.

Comment (MrAlaux) [reply src/m_menu.c:4528]: I'll wait for Fabian's input on it.

**Comment by fabiangreffrath** [comment]
Hm, okay, let's have two options for this. If we go this way, I think it makes more sense to have them formulated in a positive way, i.e. "show xy messages" instead of "hide xy messages".

**Comment by fabiangreffrath** [comment]
@MrAlaux You might wanto to merge the master branch back into your working branch in order to get the CI build error fixed.

**Comment by MrAlaux** [comment]
Alright, I'll try to do it when I get back to working on this PR.

**Comment by MrAlaux** [comment]
Well, the build worked in 62dacb7 but not in c4f7a5e, doesn't sound like I'm the one who messed up on the latter. Oh well.

**Comment by rfomin** [comment]
> Well, the build worked in [62dacb7](https://github.com/fabiangreffrath/woof/commit/62dacb709b749889bf3b703ef8ff9c08184941e6) but not in [c4f7a5e](https://github.com/fabiangreffrath/woof/commit/c4f7a5edf863dfca27682b3f32c36339f400ce78), doesn't sound like I'm the one who messed up on the latter. Oh well.

Looks like a network problem. I think we can ignore it for now.

**Review by rfomin — APPROVED** [review]
Looks good, thanks!

File: src/am_map.c [hunk 3]
```diff
@@ -955,9 +953,9 @@ boolean AM_Responder
 
       switch (automapoverlay)
       {
-        case 2:  plr->message = "Dark Overlay On";  break;
-        case 1:  plr->message = s_AMSTR_OVERLAYON;  break;
-        default: plr->message = s_AMSTR_OVERLAYOFF; break;
+        case 2:  doomprintf(MESSAGES_NONE, "Dark Overlay On");        break;
```
Comment (fabiangreffrath) [inline src/am_map.c:956]: Strcitly speaking, the second argument isn't a format string here. Better use the `"%s", string` syntac consistently.

File: src/g_game.c [hunk 4]
```diff
@@ -1124,7 +1124,7 @@ static void G_ReadDemoTiccmd(ticcmd_t *cmd)
 	  cmd->buttons & BTS_SAVEGAME)
 	{
 	  cmd->buttons &= ~BT_SPECIALMASK;
-	  players[consoleplayer].message = "Game Saved (Suppressed)";
+	  doomprintf(MESSAGES_NONE, "Game Saved (Suppressed)");
```
Comment (fabiangreffrath) [inline src/g_game.c:1127]: Same here.

File: src/g_game.c [hunk 5]
```diff
@@ -4000,10 +4000,16 @@ boolean G_CheckDemoStatus(void)
...
 #define MAX_MESSAGE_SIZE 1024
 
-void doomprintf(const char *s, ...)
+void doomprintf(int category, const char *s, ...)
 {
   static char msg[MAX_MESSAGE_SIZE];
   va_list v;
+  extern int show_toggle_messages, show_pickup_messages;
+
+  if (  (category == MESSAGES_TOGGLE && !show_toggle_messages)
```
Comment (fabiangreffrath) [inline src/g_game.c:4009]: Nitpick, but I prefer the following style:
```c
if ((a == b && !c) ||
    (b == c && !d)
{
  ...
}
```

File: src/z_zone.h [hunk 6]
```diff
@@ -56,7 +56,15 @@ void *Z_Calloc(size_t n, size_t n2, pu_tag tag, void **user);
 void *Z_Realloc(void *p, size_t n, pu_tag tag, void **user);
 
 // Doom-style printf
-void doomprintf(const char *, ...) PRINTF_ATTR(1, 2);
+
+enum {
+  MESSAGES_NONE,
+  MESSAGES_TOGGLE,
+  MESSAGES_PICKUP,
+  MESSAGES_BOTH,
```
Comment (fabiangreffrath) [inline src/z_zone.h:64]: I think we don't need `MESSAGES_BOTH` anymore.

**Review by fabiangreffrath — CHANGES_REQUESTED** [review]
Thanks for this! My only complaint: Please make sure that the second parameter to `doomprintf()` does always contain a format identifier, else we might get error messages with certain compiler flags. I have identified the first two occurences, but there are quite some more.

File: src/am_map.c (hunk 3 above)
Comment (rfomin) [reply src/am_map.c:956]: Do we have to write like this? Is `printf("string");` illegal these days?

Comment (fabiangreffrath) [reply src/am_map.c:956]: We explicitly declare it with `PRINTF_ATTR(2, 3)` and I think this "raises compiler awareness". I know for sure that its illegal to pass a char pointer in that case, maybe it's less strict with string literals.

Comment (fabiangreffrath) [reply src/am_map.c:956]: Alright. To make sure we catch all potential format string errors, I was going to add `-Wformat=2` to our compiler flags - only to find out that it is already there. So, in this case, everything that doesn't emit a warning is valid. I would appreciate, though, if you could decide for one style or the other - and probably it even makes more sense to decide for the simpler solution.

Comment (MrAlaux) [reply src/am_map.c:956]: By "one style or the other", do you mean `doomprintf(MESSAGES_NONE, "string")` vs. `doomprintf(MESSAGES_NONE, "%s", string)`?

Comment (fabiangreffrath) [reply src/am_map.c:956]: Yes. And apparently I was mistaken and the former is safe for string literals, so that should be the preferred format.

Comment (kraflab) [reply src/am_map.c:956]: Ya, the only concern is if you do something like `printf(variable)`, where you can't guarantee there aren't arguments in the string.

Comment (fabiangreffrath) [reply src/am_map.c:956]: `AMSTR_OVERLAYON` is a macro (i.e. a string literal), `s_AMSTR_OVERLAYON` is a `char *` pointing to that macro.

Comment (rfomin) [reply src/am_map.c:956]: I think you are already doing everything right. `printf("string literal");` and `printf("%s", string_var);`

Comment (fabiangreffrath) [reply src/am_map.c:956]: Then it was most probably me who mixed things up this morning, sorry for the noise!

**Comment by fabiangreffrath** [comment]
If you could fix the other two nitpicks that I mentioned earlier (and the conflict in `src/g_game.c`) then I'd consider this ready to merge.

**Comment by MrAlaux** [comment]
Sure thing, I'll get to it as soon as I can.

**Review by fabiangreffrath — APPROVED** [review]

**Comment by MrAlaux** [comment]
Oh, I was just about to merge the master branch myself... oh well!

**Comment by fabiangreffrath** [comment]
Thank you!

**Pull request merged**